semgrep{,-core}: 1.15.0 -> 1.27.0

06kellyjac opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Can no longer fetch prebuilt semgrep-core from github but can get it from the python wheel
We're still able to build the rest of semgrep from source

---

related: https://github.com/returntocorp/semgrep/issues/7570

resolves: #235417

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

06kellyjac commented:
TODO:

Fixup semgrep-core

```
λ ./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35542 files tracked by git with 1 Code rule:
  Scanning 1237 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--                                                                                                                        
[Errno 13] Permission denied: '/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/bin/semgrep-core'
Traceback (most recent call last):
  File "/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/commands/wrapper.py", line 35, in wrapper
    func(*args, **kwargs)
  File "/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/commands/scan.py", line 907, in scan
# ...
```

I thought I'd get the ball rolling on the PR so someone can help me with the fetchpypi bit.

06kellyjac commented:
Fixed, it came with a static semgrep-core but it wasn't marked executable for some reason...

```
λ ./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35542 files tracked by git with 1 Code rule:
  Scanning 1237 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00                                                                                                                        
                
                
┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.
  Scan skipped: 2 files larger than 1.0 MB, 89 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1237 files: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help
```

06kellyjac commented:
@mweinelt since you do a bit of python would you mind reviewing this & maybe helping with fetching the wheel properly :upside_down_face:

06kellyjac commented on pkgs/tools/security/semgrep/default.nix:
```diff
@@ -1,41 +1,28 @@
-    common.submodules)) + ''
-    cd cli
-  '';
+  src = fetchurl {
+    url = "https://files.pythonhosted.org/packages/33/c5/c236d76a2b7501355e4957440b15aeb6c375b58bd1adb6ed5af16585cad1/semgrep-1.24.1-cp37.cp38.cp39.py37.py38.py39-none-any.whl";
+    hash = "sha256-qpBchqKJBJcg3uHLuTnF3y9yzAb5yKI2Vm927WgzQQI=";
+  };
```
Looks like it is the same as `python` here

https://github.com/BlackArch/blackarch/blob/master/packages/semgrep/PKGBUILD#L23

06kellyjac commented:
Result of `nixpkgs-review pr 235466` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

06kellyjac commented:
@ambroisie I realised even though the wheel is "none-any" it isn't actually going to be a cross platform binary & I didn't remove the darwin platforms.

I've now changed the updater and fetching for semgrep-core to actually get the darwin copies of the wheels

reckenrode commented:
Result of `nixpkgs-review pr 235466` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

reckenrode commented:
> Thanks @reckenrode! Did you try run `semgrep`? e.g. `./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'`

This is what I get running that command.

```console
$ ./results/semgrep/bin/semgrep  scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 7 files tracked by git with 1 Code rule:
  Scanning 1 file.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌──────────────┐
│ Scan Summary │
└──────────────┘

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1 file: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help
```

reckenrode commented:
Result of `nixpkgs-review pr 235466` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

reckenrode commented:
```
$ ./results/semgrep/bin/semgrep  scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35634 files tracked by git with 1 Code rule:
  Scanning 1220 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan skipped: 2 files larger than 1.0 MB, 89 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1220 files: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help

```

06kellyjac commented:
Thanks for testing on both platforms :heart:

ambroisie commented on pkgs/tools/security/semgrep/update.sh:
```diff
@@ -50,43 +50,54 @@ fetchgithub() {
+fetch_arch() {
+  VERSION=$1
+  PLATFORM=$2
+  nix-prefetch "{ fetchPypi }:
+fetchPypi rec {
+  pname = \"semgrep\";
+  version = \"$VERSION\";
+  format = \"wheel\";
+  dist = python;
+  python = \"cp37.cp38.cp39.py37.py38.py39\";
+  platform = \"$PLATFORM\";
+}
+"
```
I believe you need to add it to the `nix-shell` list of packages?

06kellyjac replied:
you're right. added

ambroisie reviewed (approved):
Not sure about `nix-prefetch`, otherwise LGTM.

SuperSandro2000 commented on pkgs/tools/security/semgrep/semgrep-core.nix:
```diff
@@ -1,21 +1,48 @@
+    unzip
+  ];
+
+  unpackPhase = ''
+    runHook preUnpack
+    unzip $src
+    runHook postUnpack
+  '';
```
Isn't this the default when unzip is added to nativeBuildInputs?

06kellyjac replied:
Not sure. I thought it got upset because the extension is .whl

I can try again without

06kellyjac replied:
Tried it and yeah the unzip unpacker checks for a `.zip` extension.

I guess we could modify it to accept `.whl` and maybe `.ZIP` while we're at it

https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/unzip/setup-hook.sh#L3

06kellyjac replied:
For now I have added a comment

SuperSandro2000 commented on pkgs/tools/security/semgrep/default.nix:
```diff
@@ -97,10 +104,12 @@ buildPythonApplication rec {
 
   meta = common.meta // {
     description = common.meta.description + " - cli";
+    platforms = semgrep-core.meta.platforms;
```
```suggestion
    inherit (semgrep-core.meta) platforms;
```

06kellyjac committed:
semgrep{,-core}: 1.15.0 -> 1.27.0

Now fetching semgrep-core from the python wheel as r2c don't add
binaries to the GH releases anymore.

emilytrau commented:
Resolves? #188944

06kellyjac commented:
yes. I wasn't aware of that issue

SuperSandro2000 commented on pkgs/tools/security/semgrep/semgrep-core2.nix:
```diff
@@ -0,0 +1,73 @@
+{ lib, stdenv, fetchFromGitHub, dune_3, pcre, ocamlPackages }:
```
Did you commit this file by accident?

06kellyjac replied:
Yes lol. Ill remove it when im at my pc

06kellyjac replied:
removed

DamienCassou commented:
Anything preventing the merge of this PR?

DamienCassou commented:
Result of `nixpkgs-review pr 235466` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

Janik-Haag reviewed (approved):
Build's and runs fine locally. One maintainer already approved this so I will go ahead and merge this.

Janik-Haag commented:
@06kellyjac why did you add the backport label?

06kellyjac commented:
Why not?

Janik-Haag commented:
Because we (at least as far as I know) only backport secrity and bug fixes but don't back port version updates.

DamienCassou commented:
Here is what the manual as to say about backports: https://nixos.org/manual/nixpkgs/stable/#acceptable-backport-criteria.

06kellyjac commented:
Thanks for linking @DamienCassou 

So yes, this falls under

> Version updates which include new functionality (but no breaking changes) 

Most people wont add the backport label because its extra effort, but in this case semgrep was stuck on 1.15.0 for ages and there was a version update request issue.
If it isnt backported 23.05 will have the same old version of semgrep as 22.11 all the way until 23.11

Janik-Haag commented:
Okay thanks for explaining, can you test and approve the backport?
